### PR TITLE
use optionnal in all method findById

### PR DESCRIPTION
**Pull request by Donatien26**

### CODE REVIEW

File: sugoi-api-core/src/main/java/fr/insee/sugoi/core/service/ApplicationService.java [hunk 1]
```diff
@@ -16,32 +16,38 @@
...
 import fr.insee.sugoi.core.model.PageableResult;
 import fr.insee.sugoi.model.Application;
+import java.util.Optional;
 
 public interface ApplicationService {
 
   /**
-   * Creates an application in the realm
+   * Verify if the application already exist in the realm (by it's name) and create if it does'nt
+   * exist
```
Comment (CChemin) [inline sugoi-api-core/src/main/java/fr/insee/sugoi/core/service/ApplicationService.java:25]: doesn't

File: sugoi-api-rest-old-services/src/main/java/fr/insee/sugoi/old/services/controller/IdController.java [hunk 2]
```diff
@@ -94,7 +95,7 @@ private Contact findContact(Realm realm, String id) {
     try {
       return ouganextSugoiMapper.serializeToOuganext(
           userService.findById(realm.getName(), null, id), Contact.class);
-    } catch (EntityNotFoundException e) {
+    } catch (UserNotFoundException e) {
```
Comment (CChemin) [inline sugoi-api-rest-old-services/src/main/java/fr/insee/sugoi/old/services/controller/IdController.java:98]: findById doesn't throw UserNotFoundException does it ? here we should remove the try and check if we have something in the optional

File: sugoi-api-rest-services/src/main/java/fr/insee/sugoi/services/controller/GroupController.java [hunk 3]
```diff
@@ -144,22 +146,24 @@
       @Parameter(description = "Name of application where to add group", required = true)
           @PathVariable(value = "application", required = true)
           String applicationName,
-      @Parameter(description = "Group to create", required = true) @RequestBody Group group) {
+      @Parameter(description = "Group to create", required = true) @RequestBody
+          GroupView groupView) {
```
Comment (CChemin) [inline sugoi-api-rest-services/src/main/java/fr/insee/sugoi/services/controller/GroupController.java:150]: What is a groupView ? maybe specify what is expected in swagger doc

File: sugoi-api-core/src/main/java/fr/insee/sugoi/core/service/ApplicationService.java [hunk 4]
```diff
@@ -16,32 +16,38 @@
...
    * @param application
    * @return The created application
+   * @throws ApplicationAlreadyExistException if application already exist in realm
    */
   Application create(String realm, Application application);
 
   /**
-   * Updates application, if the application has groups, they are also updated. Warning, this is not
-   * an atomic process (an app can be updated and not the groups)
+   * if application exist, updates application, and if the application has groups, they are also
```
Comment (micedre) [inline sugoi-api-core/src/main/java/fr/insee/sugoi/core/service/ApplicationService.java:35]: *exists

File: sugoi-api-core/src/main/java/fr/insee/sugoi/core/service/GroupService.java [hunk 5]
```diff
@@ -13,24 +13,90 @@
...
 public interface GroupService {
 
+  /**
+   * Check if the group exist in the app (by name) and create it if it doesn't exist
+   *
+   * @param realm
+   * @param appName
+   * @param group
+   * @return the created group
+   * @throws GroupAlreadyExistException if the group already exist
```
Comment (micedre) [inline sugoi-api-core/src/main/java/fr/insee/sugoi/core/service/GroupService.java:33]: How do we know if the group exist (by name ? is it case sensitive?)

File: sugoi-api-core/src/main/java/fr/insee/sugoi/core/service/impl/ApplicationServiceImpl.java [hunk 6]
```diff
@@ -34,58 +37,74 @@
...
-        Map.ofEntries(Map.entry(EventKeysConfig.APPLICATION, application)));
-    return storeProvider.getWriterStore(realm).createApplication(application);
+    if (!findById(realm, application.getName()).isPresent()) {
+      String appName = storeProvider.getWriterStore(realm).createApplication(application).getName();
+      sugoiEventPublisher.publishCustomEvent(
+          realm,
+          null,
+          SugoiEventTypeEnum.CREATE_APPLICATION,
+          Map.ofEntries(Map.entry(EventKeysConfig.APPLICATION, application)));
+      return findById(realm, appName)
```
Comment (micedre) [inline sugoi-api-core/src/main/java/fr/insee/sugoi/core/service/impl/ApplicationServiceImpl.java:47]: I find it weird to throw an ApplicationNotFoundException here. If the app does not exist, this means we could not create it, it should have thrown another exception I think.

File: sugoi-api-core/src/main/java/fr/insee/sugoi/core/service/GroupService.java (hunk 5 above)
Comment (Donatien26) [reply sugoi-api-core/src/main/java/fr/insee/sugoi/core/service/GroupService.java:33]: for a group it's not case sensitive but for a user it depends of the implementation (file different from ldap)

File: sugoi-api-core/src/main/java/fr/insee/sugoi/core/exceptions/OrganizationNotFoundException.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+package fr.insee.sugoi.core.exceptions;
+
+public class OrganizationNotFoundException extends RuntimeException {
+
+  /** */
```
Comment (micedre) [inline sugoi-api-core/src/main/java/fr/insee/sugoi/core/exceptions/OrganizationNotFoundException.java:18]: useless

**Review by micedre — COMMENTED** [review]
with those modifications, we lose the events when  an error occured, maybe we can try to catch the exceptions in userservice an rethrow it, which would allow to publish a nex event (like CREATE_USER_ERROR or a new property `error` on the existing events name) ? What do you think ?

**Review by micedre — APPROVED** [review]

**Pull request merged**